### ISSUE TITLE
Nats_Engine: Pos in empty own buffer: While executing NATS.

### ISSUE THREAD

**Issue by Kebastos** [ClickHouse/ClickHouse#40062]
> You have to provide the following information whenever possible.

When using Nats_Engine, with increasing load, errors appear in the logs and message processing stops.

During local testing, an error appears from time to time, when the load is in production, the error appears immediately at the start, not a single message has time to be processed.

How to reproduce:

```
  CREATE TABLE queue_prod (
    message String
  ) ENGINE = NATS
    SETTINGS nats_url = '10.193.68.253', --'172.18.0.1:4222',
             nats_subjects = 'Metrics',
             nats_format = 'JSONAsString',
             nats_username = 'user',
             nats_password = 'password';
```

```
  CREATE TABLE messages_prod (
    message String
) ENGINE = MergeTree() ORDER BY message;
```

```
  CREATE MATERIALIZED VIEW consumer_prod TO messages_prod
    AS SELECT
            message
       FROM queue_prod;
```
And then send more than 500 messages per minute.

Expected behavior: 
All messages have been processed.

Error message and/or stacktrace:
```
2022.08.10 11:13:30.198289 [ 262 ] {} <Error> void DB::StorageNATS::streamingToViewsFunc(): Code: 49. DB::Exception: Pos in empty own buffer: While executing NATS. (LOGICAL_ERROR), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xba37dda in /usr/bin/clickhouse
1. DB::PeekableReadBuffer::checkStateCorrect() const @ 0xbab722e in /usr/bin/clickhouse
2. DB::PeekableReadBuffer::nextImpl() @ 0xbab7caf in /usr/bin/clickhouse
3. DB::IInputFormat::resetParser() @ 0x17bb6a79 in /usr/bin/clickhouse
4. DB::JSONAsRowInputFormat::resetParser() @ 0x17c56e2d in /usr/bin/clickhouse
5. DB::StreamingFormatExecutor::execute() @ 0x17bb65ce in /usr/bin/clickhouse
6. DB::NATSSource::generate() @ 0x171015bd in /usr/bin/clickhouse
7. DB::ISource::tryGenerate() @ 0x17b8fa95 in /usr/bin/clickhouse
8. DB::ISource::work() @ 0x17b8f5e6 in /usr/bin/clickhouse
9. DB::ExecutionThreadContext::executeTask() @ 0x17babe0a in /usr/bin/clickhouse
10. DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) @ 0x17ba0cf0 in /usr/bin/clickhouse
11. DB::PipelineExecutor::executeImpl(unsigned long) @ 0x17b9fd64 in /usr/bin/clickhouse
12. DB::PipelineExecutor::execute(unsigned long) @ 0x17b9fafd in /usr/bin/clickhouse
13. DB::CompletedPipelineExecutor::execute() @ 0x17b9e76c in /usr/bin/clickhouse
14. DB::StorageNATS::streamToViews() @ 0x170f4513 in /usr/bin/clickhouse
15. DB::StorageNATS::streamingToViewsFunc() @ 0x170f2f88 in /usr/bin/clickhouse
16. DB::BackgroundSchedulePoolTaskInfo::execute() @ 0x15e2a2dd in /usr/bin/clickhouse
17. DB::BackgroundSchedulePool::threadFunction() @ 0x15e2d3d6 in /usr/bin/clickhouse
18. ? @ 0x15e2e28e in /usr/bin/clickhouse
19. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xbb046e8 in /usr/bin/clickhouse
20. ? @ 0xbb07a7d in /usr/bin/clickhouse
21. ? @ 0x7fb872055609 in ?
22. __clone @ 0x7fb871f7a133 in ?
 (version 22.7.2.15 (official build))
```

Clickhouse version: 22.7.2.15

**Comment by Kebastos** [comment]
Hello
Will there be any information on this problem?
Do you need any more information?

**Comment by tanleach** [comment]
Also seeing the same issue:
`2023.01.17 15:34:09.395827 [ 278 ] {} <Error> void DB::StorageNATS::streamingToViewsFunc(): Code: 49. DB::Exception: Pos in empty own buffer: While executing NATS. (LOGICAL_ERROR), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked const&, int, bool) @ 0xc809de8 in /usr/bin/clickhouse
1. ? @ 0x72537bc in /usr/bin/clickhouse
2. DB::PeekableReadBuffer::checkStateCorrect() const @ 0xc85f5bc in /usr/bin/clickhouse
3. DB::PeekableReadBuffer::nextImpl() @ 0xc860078 in /usr/bin/clickhouse
4. DB::IInputFormat::resetParser() @ 0x1241059c in /usr/bin/clickhouse
5. DB::JSONAsRowInputFormat::resetParser() @ 0x124b8278 in /usr/bin/clickhouse
6. DB::StreamingFormatExecutor::execute() @ 0x124100c4 in /usr/bin/clickhouse
7. DB::NATSSource::generate() @ 0x11e0a484 in /usr/bin/clickhouse
8. DB::ISource::tryGenerate() @ 0x123ecfec in /usr/bin/clickhouse
9. DB::ISource::work() @ 0x123ecaf0 in /usr/bin/clickhouse
10. DB::ExecutionThreadContext::executeTask() @ 0x124048e0 in /usr/bin/clickhouse
11. DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) @ 0x123fa74c in /usr/bin/clickhouse
12. DB::PipelineExecutor::executeImpl(unsigned long) @ 0x123f99b4 in /usr/bin/clickhouse
13. DB::PipelineExecutor::execute(unsigned long) @ 0x123f9794 in /usr/bin/clickhouse
14. DB::CompletedPipelineExecutor::execute() @ 0x123f8234 in /usr/bin/clickhouse
15. DB::StorageNATS::streamToViews() @ 0x11dff7ac in /usr/bin/clickhouse
16. DB::StorageNATS::streamingToViewsFunc() @ 0x11dfdd6c in /usr/bin/clickhouse
17. DB::BackgroundSchedulePoolTaskInfo::execute() @ 0x108fa55c in /usr/bin/clickhouse
18. DB::BackgroundSchedulePool::threadFunction() @ 0x108fd508 in /usr/bin/clickhouse
19. ? @ 0x108fe378 in /usr/bin/clickhouse
20. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xc8ade48 in /usr/bin/clickhouse
21. ? @ 0xc8b29a8 in /usr/bin/clickhouse
22. start_thread @ 0x7624 in /usr/lib/aarch64-linux-gnu/libpthread-2.31.so
23. ? @ 0xd149c in /usr/lib/aarch64-linux-gnu/libc-2.31.so
 (version 22.12.3.5 (official build))`

**Comment by kssenii** [comment]
I think it could be already fixed by https://github.com/ClickHouse/ClickHouse/pull/42777. Could you try the master version?
master version can be installed this way
```
curl -O 'https://builds.clickhouse.com/master/amd64/clickhouse' && chmod a+x clickhouse
sudo ./clickhouse install
```
@Avogar, what do you think?

**Comment by kssenii** [comment]
hm, I see the same issue https://github.com/ClickHouse/ClickHouse/issues/38061 but for RabbitMQ, and it was fixed (then the fix for Nats should be the same), then https://github.com/ClickHouse/ClickHouse/pull/42777 probably does not fix.

**Comment by kssenii** [comment]
oh, I found the fix https://github.com/ClickHouse/ClickHouse/pull/42777/files#diff-036296491b4ac457a2ae7d91fec86d5f9a8afc87109bf229d80fb2f021f8839dL106, indeed this issue is fixed in https://github.com/ClickHouse/ClickHouse/pull/42777.

**Comment by tanleach** [comment]
I am using the Docker image with the latest tag. is that not the master version? (On M1 Mac, fwiw)

**Comment by Avogar** [comment]
Changes from that PR weren't included in any release yet. It will be available since 23.1.
Maybe we need to think about backporting the fix manually.

**Comment by kssenii** [comment]
>I am using the Docker image with the latest tag. is that not the master version? 

No, it is not the master version.
